pkg/apis/build/v1alpha1/buildstrategy_types.go

```diff
@@ -54,3 +51,26 @@ type BuildStrategyList struct {
 func init() {
 	SchemeBuilder.Register(&BuildStrategy{}, &BuildStrategyList{})
 }
+
+// StrategyRef can be used to refer to a specific instance of a buildstrategy.
+// Copied from CrossVersionObjectReference: https://github.com/kubernetes/kubernetes/blob/169df7434155cbbc22f1532cba8e0a9588e29ad8/pkg/apis/autoscaling/types.go#L64
+type StrategyRef struct {
+	// Name of the referent; More info: http://kubernetes.io/docs/user-guide/identifiers#names
+	Name string `json:"name,omitempty"`
+	// BuildStrategyKind indicates the kind of the buildstrategy, namespaced or cluster scoped.
+	Kind BuildStrategyKind `json:"kind,omitempty"`
```

Comment:
Could we mark only `Name` as mandatory and everything else as optional ? If only Name is specified by the `Build`, the code could assume that the user is expecting a namespace-scoped BuildStrategy.


For all optional fields: 

May I request the following change, please?
```
// BuildStrategyKind indicates the kind of the buildstrategy, namespaced or cluster scoped.
// +optional
Kind *BuildStrategyKind `json:"kind,omitempty"`
```

The reason I prefer using pointers is that it helps our code distinguish between the following:

- If the user has specified an empty field
- if the user has skipped the field


`*BuildStrategyKind` is used:
- when nil shows up in our controller after serialization, we know that the field was skipped in the CR.
- when "" shows up in our controller after serialization, we know that an empty string was passed in the CR.


`BuildStrategyKind` is used:
When "" shows up in our controller after serialization, we know the field might have **either** been skipped in the CR, **Or** the user might have specified an empty list in the CR <-- this is an ambiguity.

To avoid this ambiguity, I prefer using pointers for optional fields because it lets us write code to handle scenarios as applicable.

Reply:
Yep, I agree with this change.

Let me confirm the results of `nil` and `""`

Do you mean:
- if `*BuildStrategyKind` is `nil`, it means the user doesn't set it, we should set the default `BuildStrategy` for him
- if `*BuildStrategyKind` is `""`, it means the user set it as empty BuildStrategyKind, we should report an error and tell him that he input a wrong BuildStrategyKind.

Did I understand correctly?

Reply:
Correct.

Reply:
Hi @sbose78 ,

I have refined the StrategyRef in this commit:
https://github.com/redhat-developer/build/pull/62/commits/cf42ef58aa0bf3de2cca8373246d77b8b05bc704

I didn't improve the build-controller unit test in this PR.

I think after we confirm or developing the Build/BuildRun implementation, we should use the operator-sdk `fake client` to test the controller locally better. So we should do in that PR.

And I have verified in my env, it works fine. Pls review.

Reply:
Yeah, makes sense!